app/src/org/commcare/android/database/global/GlobalDatabaseUpgrader.java

```diff
@@ -77,4 +90,40 @@ private boolean upgradeTwoThree(SQLiteDatabase db) {
             db.endTransaction();
         }
     }
+
+    /**
+     * Prior to multiple application seating, the FormsProvider and the InstanceProvider were both
+     * using one global database for all forms/instances. Now that we can have multiple apps
+     * installed at once, we need to have one forms db and one instances db per app. This method
+     * performs the necessary one-time migration from a global db that still exists on a device
+     * to the new per-app system
+     */
+    private boolean upgradeProviderDb(SQLiteDatabase db, ProviderUtils.ProviderType type) {
+        File oldDbFile = CommCareApplication._().getDatabasePath(type.getOldDbName());
+        if (oldDbFile.exists()) {
+            File newDbFile = CommCareApplication._().getDatabasePath(
+                    ProviderUtils.getProviderDbName(type, getInstalledAppRecord(c, db).getApplicationId()));
+            if (!oldDbFile.renameTo(newDbFile)) {
+                // Big problem, should potentially crash here ?
+                return false;
+            } else {
+                return true;
+            }
+        }
+        return true;
+    }
+
+    private static ApplicationRecord getInstalledAppRecord(Context c, SQLiteDatabase db) {
+        SqlStorage<Persistable> storage = new SqlStorage<Persistable>(
+                ApplicationRecord.STORAGE_KEY,
+                ApplicationRecord.class,
+                new ConcreteDbHelper(c, db));
+        for (Persistable p : storage) {
+            ApplicationRecord r = (ApplicationRecord) p;
+            if (r.getStatus() == ApplicationRecord.STATUS_INSTALLED) {
```

Comment:
If there's more than one installed record we likely should either deliberately crash or print a message saying "You need to wipe and restart CommCare" if it's easy enough to do so.